### ISSUE TITLE
400 Invalid Attachment Warning / Error in Pull Replication

### ISSUE THREAD

**Issue by davidquon** [couchbase/couchbase-lite-ios#961]
We were previously using iOS CBL `1.0.3.1` and recently upgraded to iOS CBL `1.1.1`.  Now with certain documents with attachments we started seeing errors like this which stop the replication.

```
2015-10-24 11:21:20.188 AppName[6574:1993505] WARNING: CBL_Puller[https://username:*****@hostname:6984/database] failed to write {namespace/doc_id #198-b707c081c57791deb57a366aee25c141}: status=491
2015-10-24 11:21:20.189 AppName[6574:1993505] Sync: CBL_Puller[https://username:*****@hostname:6984/database] Progress: set error = 400 Invalid attachment
2015-10-24 11:21:20.190 AppName[6574:1993505] Sync: CBL_Puller[https://username:*****@hostname:6984/database] STOPPING...
```

We are running CouchDB version `1.6.1` for our couch server.  I'm reproducing this problem with a completely new CBL database each time.

Now if you query the CouchDB for `namespace/doc_id` I get:
```
"_id":"namespace/doc_id","_rev":"366-8fc5ffd358688535b06788b5872793bc”,
```

If I query CouchDB for `namespace/doc_id` with `?revs_info=true` I get back only 10 revisions as it was configured.

So why is CBL complaining about problems with `kCBLStatusBadAttachment` for revisions like 75-100 revisions prior to the latest?  Also when using CBL `1.0.3.1` with the same exact database this issue doesn't occur and everything replicates as expected.  Any suggestions of how to solve this problem would be greatly appreciated as so far none of the workarounds I've tried have yielded a solution.  Thanks.

This issue was created off the Google group discussion.
https://groups.google.com/forum/#!topic/mobile-couchbase/MHrzXr1aWxw

**Comment by davidquon** [comment]
> That revision must be in the _changes feed on the server, since that’s where the puller finds out what to download.

This helped lead us to some clues of what CBL is having trouble with.  Looking at the changes feed on CouchDB it seems this document's changes feed is not in order.

```
{"seq":5382,"id":"namespace/doc_id","changes":[
{"rev":"366-8fc5ffd358688535b06788b5872793bc"},
{"rev":"345-adf52b7bdce442640133ad4a3ae27782"},
{"rev":"330-60b08f47f1da203c8346b93b930abb01"},
{"rev":"296-86856304afbefb3375dba08d6364783c"},
{"rev":"216-dca59ef3e2347150011ac2113f71006c"},
{"rev":"198-b707c081c57791deb57a366aee25c141"},
{"rev":"194-8970209f197a2138da0d7e79f1c59e3e"},
{"rev":"184-4f11477796598c7fd3d41e36257befcb"},
{"rev":"182-080145b2b0ecc7d037fa0da4dcaa4b1d"},
{"rev":"167-1bdc1755f6c1c1b3f95494eb04b1f060"},
{"rev":"134-53ff2a7eabb5bba9e17419dc94aad08b"},
{"rev":"131-c3b5f5ac7c316211db1a39512d244823"},
{"rev":"91-8e48e1b7c1aa16821c785e6235dad78c"},
{"rev":"73-63d4762c4357364abede2a042eb4e062"},
{"rev":"67-bb5421c01de5ccaf783d5ad6d3abb4cf"},
{"rev":"65-6f13ae182f6ba0dc02b6f197916fd1ac"},
{"rev":"54-3d09e0a86ac79cebd70e054a31b02c97"},
{"rev":"41-e967be5393174a7d2ccbb85d3ac7c847"},
{"rev":"199-5debfc11db6448e40bb411f961c6f868"},
{"rev":"195-d108aca73cd20e785b4e87278478171f"},
{"rev":"185-94f598b82a4dc4ef0ac6338a84e1a32b"},
{"rev":"183-2ba1f30207b9552f029479ec179c88e1"},
{"rev":"168-84dbaad6f19c6e84fe6eeb1362dccd7e"},
{"rev":"135-2b4afe411605d9a0437adb71513e5526"},
{"rev":"132-02f0c247ba5f85e3ef7e6838130dbb3f"},
{"rev":"92-c76fd07e24f084b9b8d4816af60d1814"},
{"rev":"74-4bf733e309307c0e616bdaf2c39d5832"},
{"rev":"68-9af36a388eb16ef1fb20fd2b6c1dc8d6"},
{"rev":"66-9ee62596bf2558f3976f59b296aaee23"},
{"rev":"55-eec8e04c84b74d3e54d115096ff4806b"},
{"rev":"42-f21248179c06cd8de2d30661cf1242fd"}]},
```

This is the block of conflicts for the same document:
```
"_conflicts":[
"345-adf52b7bdce442640133ad4a3ae27782",
"330-60b08f47f1da203c8346b93b930abb01",
"296-86856304afbefb3375dba08d6364783c",
"216-dca59ef3e2347150011ac2113f71006c",
"198-b707c081c57791deb57a366aee25c141",
"194-8970209f197a2138da0d7e79f1c59e3e",
"184-4f11477796598c7fd3d41e36257befcb",
"182-080145b2b0ecc7d037fa0da4dcaa4b1d",
"167-1bdc1755f6c1c1b3f95494eb04b1f060",
"134-53ff2a7eabb5bba9e17419dc94aad08b",
"131-c3b5f5ac7c316211db1a39512d244823",
"91-8e48e1b7c1aa16821c785e6235dad78c",
"73-63d4762c4357364abede2a042eb4e062",
"67-bb5421c01de5ccaf783d5ad6d3abb4cf",
"65-6f13ae182f6ba0dc02b6f197916fd1ac",
"54-3d09e0a86ac79cebd70e054a31b02c97",
"41-e967be5393174a7d2ccbb85d3ac7c847”]
```

If we purge all of these conflicts on the CouchDB side then the pull replication is successful if it's into a new CBL database.

**Comment by davidquon** [comment]
If we have the above document with conflicts already in a CBL `1.0.3.1` database and attempt to upgrade to a CBL `1.1.1` database we'll get the following exception.  Not sure if this is the same root cause as above or warrants a separate issue.
```
2015-10-26 17:34:37.041 AppName[6767:2078444] CouchbaseLite: Upgrading database v1.0 (15) to v1.1 at /var/mobile/Containers/Data/Application/FAA81CC9-6E58-4121-86BE-E8FF6D58A0A7/Library/Application Support/CouchbaseLite/database.cblite ...
2015-10-26 17:34:37.952 AppName[6767:2078444] WARNING: Upgrade failed: status 491
2015-10-26 17:34:38.148 AppName[6767:2078266] WARNING: CBLDatabase: Database version (15) is older than I know how to work with
libc++abi.dylib: terminate_handler unexpectedly threw an exception
```

**Comment by davidquon** [comment]
If we purge all of these conflicts on the CouchDB side, replicate to CBL, and try to upgrade from a CBL `1.0.3.1` database to a CBL `1.1.1` we get the same exception above despite all of the conflicts being purged.

**Comment by snej** [comment]
Any chance you could send us a zipped copy of the pre-upgrade 1.0.3 database? Then we could open it up in 1.1.1 and reproduce the error. In my experience, these 491-Bad-Attachment issues are hard to troubleshoot until we can step through them in the debugger.

(If the db contains sensitive information, you can email it to me directly, or upload it somewhere private and email me a link. `jens` at couchbase. I'll sign an NDA if necessary but it'll add a day or so of latency.)

**Comment by davidquon** [comment]
Hi Jens.  Since this is unfortunately a live database, I emailed you an NDA.  Let me know if you didn't get it or have any questions.  I'll zip up a copy of the pre-upgrade database for you.  We can also try to setup a database for you to replicate from using CBL `1.1.1` unless you think what you find with the pre-upgrade DB is the same root cause as in the replicator.  Thanks for your help.

**Comment by davidquon** [comment]
I tried with the current master `01ab84b` and still have the same problem.  However now there is another trace.
```
2015-10-30 11:07:52.106 AppName[37255:1000827] CouchbaseLite: Upgrading database v1.0 (16) to v1.1 at /Users/user/Library/Developer/CoreSimulator/Devices/AC156564-A80E-484B-924E-2673BBF0D21C/data/Containers/Data/Application/BFD9E66C-5B98-48E9-8F4F-A633758E5245/Library/Application Support/CouchbaseLite/database.cblite ...
11:07:52.121| WARNING: CBLDatabase: No pending attachment for digest sha1-wfz47ZXUaFU0VargEILAnCtCOs0=
11:07:52.122| WARNING: Upgrade failed: status 491
```

**Comment by snej** [comment]
I got the NDA and will get it back to you this afternoon.

**Comment by davidquon** [comment]
@snej - Just a friendly ping about the NDA so I can email you the zipped database.  Thanks for all your help Jens.

**Comment by snej** [comment]
Sorry! It got lost in my inbox. I've signed it now and emailed you the PDF.

**Comment by davidquon** [comment]
No worries @snej.  Thanks for the signed NDA.

I emailed you the CBL database built with CBL 1.0.x.  If opened with CBL 1.1.x (or master) it will fail during the upgrade with "Upgrade failed: status 491".

We have a similar problem when replicating this database from CouchDB to CBL 1.1.x whereas CBL 1.0.x works fine.

After investigating the upgrade can you please let us know if you think they are the same root cause and if there is anything we can do on the CouchDB side to remedy the problem.  We can setup a database to replicate from if it's needed for debugging.

Let me know if there is anything else I can do to help.  Thank you for all of your help and support @snej.  It's greatly appreciated.

**Comment by snej** [comment]
This is a bug in CBL's attachment-insertion code, that manifests when upgrading a database that has leaf revisions with no ancestors. This situation happens when old conflicts get left around: the leaf itself won't get pruned, but its ancestors will.

I've fixed the bug, and the upgrade works now. The only workaround I can think of is to resolve conflicts before upgrading the database, at least conflicts that are old enough that they have no known parent revision.

**Comment by snej** [comment]
Fix is on the release/1.2 branch.

**Comment by davidquon** [comment]
> The only workaround I can think of is to resolve conflicts before upgrading the database, at least conflicts that are old enough that they have no known parent revision.

We did try to resolve the conflicts on the CouchDB side and that allowed the replication to happen successfully on a new CBL 1.1.1 database.  However if the CBL 1.0.x database with conflicts already existed on the device, we resolved the conflicts on the CouchDB side, replicated these fixes into CBL 1.0.x database, and then upgraded to CBL 1.1.1 the same error would occur.

Our short term fix was to roll back to CBL `1.0.4` for the time being.  We'll test out your fix above and let you know if we have any other trouble.  Thanks for your help @snej.  It's greatly appreciated as always.
